ludwig/data/preprocessing.py

```diff
@@ -1396,6 +1389,28 @@ def embed_fixed_features(
     return results
 
 
+def _get_sampled_dataset_df(dataset_df, df_engine, sample_ratio, sample_cap, random_seed):
+    if sample_ratio < 1.0:
+        if not df_engine.partitioned and len(dataset_df) * sample_ratio < 1:
+            raise ValueError(
+                f"sample_ratio {sample_ratio} is too small for dataset of length {len(dataset_df)}. "
+                f"Please increase sample_ratio or use a larger dataset."
+            )
+
+        logger.debug(f"sample {sample_ratio} of data")
+        dataset_df = dataset_df.sample(frac=sample_ratio, random_state=random_seed)
+
+    if sample_cap:
+        if sample_cap < len(dataset_df):
```

Comment:
`len(dataset_df)` is a very expensive op for Dask dataframes (this is why we have an explicit check above to skip calling this when `df_engine.partitioned`), so calling it twice is quick succession is definitely not ideal. Let's do this instead:

```
df_len = len(dataset_df)
if sample_cap < df_len:
  # Cannot use 'n' parameter when using dask DataFrames -- only 'frac' is supported
  sample_ratio = sample_cap / df_len
  dataset_df = dataset_df.sample(frac=sample_ratio, random_state=random_seed)
```